paddle/framework/executor.cc

```diff
@@ -33,33 +33,29 @@ namespace framework {
 const std::string kFeedOpType = "feed";
 const std::string kFetchOpType = "fetch";
 
-Executor::Executor(const std::vector<platform::Place>& places) : own_(true) {
-  PADDLE_ENFORCE_GT(places.size(), 0);
-  device_contexts_.resize(places.size());
-  for (size_t i = 0; i < places.size(); i++) {
-    if (platform::is_cpu_place(places[i])) {
-      device_contexts_[i] = new platform::CPUDeviceContext(
-          boost::get<platform::CPUPlace>(places[i]));
-    } else if (platform::is_gpu_place(places[i])) {
-#ifdef PADDLE_WITH_CUDA
-      device_contexts_[i] = new platform::CUDADeviceContext(
-          boost::get<platform::GPUPlace>(places[i]));
-#else
-      PADDLE_THROW(
-          "'GPUPlace' is not supported, Please re-compile with WITH_GPU "
-          "option");
-#endif
-    }
-  }
+Executor::Executor(const std::vector<platform::Place>& places) {
+  DeviceContextPool& pool = DeviceContextPool::Get();
+  auto borrowed_contexts = pool.Borrow(places);
+  device_contexts_.swap(borrowed_contexts);
 }
 
-Executor::~Executor() {
-  if (own_) {
-    for (auto& device_context : device_contexts_) {
-      delete device_context;
-    }
-  }
-}
+// Executor::Executor(const platform::DeviceContext& device) {
+//   std::vector<const platform::Place> v = {device.GetPlace()};
+//   Executor(v);
+// }
+
+// Executor::Executor(const platform::Place& place) {
+//   std::vector<const platform::Place> v = {place};
+//   Executor(v);
+// }
+
+// Executor::~Executor() {
+//   if (own_) {
+//     for (auto& device_context : device_contexts_) {
+//       delete device_context;
+//     }
+//   }
+// }
```

Comment:
We need `Executor::Executor(const platform::DeviceContext& device)` for `recurrent_op`.
https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/operators/recurrent_op.cc#L236

Reply:
I think `Executor::Executor(const platform::DeviceContext& device)` should be removed in whole project.

Reply:
Since `Executor(places)` no longer owns the `DeviceContext`, I believe you are right.